### PR TITLE
Add NIST API and NIST test 

### PR DESCRIPTION
**Pull request by ducnguyen-aq**
In this pull request: 

- Add NIST top level API
- Specify API level via define macro in `nist_params.h`
- Add tests and verify test run correctly

Error Return: 

- Return `OQS_ERROR` and `OQS_SUCCESS` according to value defined in `oqs/common.h`

### CODE REVIEW

**Comment by ducnguyen-aq** [comment]
@kj3moraes  can you take a look?

File: Makefile [hunk 1]
```diff
@@ -2,13 +2,13 @@ CC = /usr/bin/gcc
 CFLAGS = -Wall -g -O3 -Wextra -Wpedantic
 LDLIBS = -loqs
 
-SOURCES = params.c hash.c fips202.c hash_address.c randombytes.c wots.c xmss.c xmss_core.c xmss_commons.c utils.c
-HEADERS = params.h hash.h fips202.h hash_address.h randombytes.h wots.h xmss.h xmss_core.h xmss_commons.h utils.h
+SOURCES = params.c hash.c fips202.c hash_address.c randombytes.c wots.c xmss.c xmss_core.c xmss_commons.c utils.c nist.c
```
Comment (lfousseaq) [inline Makefile:5]: Does the order of the files matter?

File: api.h [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+ * Returns 0 (success), -1 otherwise
+ **************************************************/
+int crypto_sign_keypair(unsigned char *pk, unsigned char *sk);
+
+/*************************************************
+ * Name:        XMSS_crypto_sign
+ *
+ * Description: Computes signature.
+ *
+ * Arguments:   - uint8_t *sm:   pointer to output signature (of length CRYPTO_BYTES)
```
Comment (lfousseaq) [inline api.h:23]: Please have the order of the arguments in the documentation match the order of the arguments in the function signature.

File: api.h [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+ * Name:        XMSS_crypto_remain_signatures
+ *
+ * Description: Return number of signature left
+ *
+ * Arguments:   - unsigned long long *remain: remaining signatures
+ *              - const uint8_t *sk: pointer to bit-packed private key
+ *
+ * Returns 0 (sucess), -1 otherwise
+ **************************************************/
+int crypto_remain_signatures(unsigned long long *remain, const unsigned char *sk);
```
Comment (lfousseaq) [inline api.h:61]: Does this function name come from somewhere? For me "remain_signatures" does not work grammatically. Compare with "crypto_signatures_remaining" or "crypto_remaining_signatures" or "crypto_signatures_left".

File: nist.c [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+ * Name:        crypto_sign_keypair
+ *
+ * Description: Generates public and private key.
+ *
+ * Arguments:   - uint8_t *pk: pointer to output public key (allocated
+ *                             array of LMS_CRYPTO_PUBLICKEYBYTES bytes)
+ *              - uint8_t *sk: pointer to output private key (allocated
+ *                             array of LMS_CRYPTO_SECRETKEYBYTES bytes)
+ *
+ * Returns 0 (success), -1 otherwise
```
Comment (lfousseaq) [inline nist.c:18]: Are the output arrays left unchanged on error? Is it specified?

Comment (lfousseaq) [reply nist.c:18]: Also I don't think it's a good idea to duplicate the documentation between the .h and the .c files.

File: api.h [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+ *
+ * Arguments:   - uint8_t *sm:   pointer to output signature (of length CRYPTO_BYTES)
+ *              - uint8_t *m:     pointer to message to be signed
+ *              - uint8_t *sk:    pointer to bit-packed secret key
+ *              - unsigned long long *smlen: pointer to output length of signature
+ *              - unsigned long long mlen:    length of message
+ *
+ * Returns 0 (success), -1 otherwise
+ **************************************************/
+int crypto_sign(unsigned char *sm, unsigned long long *smlen,
```
Comment (lfousseaq) [inline api.h:31]: Is the output array left unchanged on error?

File: test/nist_test.c [hunk 6]
```diff
@@ -0,0 +1,263 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <time.h>
+#include <string.h>
+#include "../api.h"
+#include "../nist_params.h"
+#include "../xmss.h"
+#include "../params.h"
+
+#define XMSS_SIGNATURES 64
```
Comment (lfousseaq) [inline test/nist_test.c:10]: Could you add a comment for what this represents?

File: xmss.h [hunk 7]
```diff
@@ -58,4 +64,11 @@ int xmssmt_sign(unsigned char *sk,
 int xmssmt_sign_open(unsigned char *m, unsigned long long *mlen,
                      const unsigned char *sm, unsigned long long smlen,
                      const unsigned char *pk);
+
+/* 
+ * Returns:
+ * remaining signatures
```
Comment (lfousseaq) [inline xmss.h:70]: nit: "the number of" remaining signatures.

File: Makefile (hunk 1 above)
Comment (ducnguyen-aq) [reply Makefile:5]: Yes it is, both `nist.c` and `nist_params.h` are on top of other files, so I put them at last.

File: api.h (hunk 5 above)
Comment (ducnguyen-aq) [reply api.h:31]: On error, the output arrays probably are modified. But I expect the caller will check error return before using those array.

File: nist.c (hunk 4 above)
Comment (ducnguyen-aq) [reply nist.c:18]: I clean up documentation in .c file

File: test/nist_test.c (hunk 6 above)
Comment (ducnguyen-aq) [reply test/nist_test.c:10]: Yes, I just add the comment

File: xmss.h (hunk 7 above)
Comment (ducnguyen-aq) [reply xmss.h]: I update the description to `The number of remaining signatures`

**Review by kj3moraes — APPROVED** [review]

**Pull request merged**